src/coreComponents/meshUtilities/InternalWellboreGenerator.cpp

```diff
@@ -0,0 +1,216 @@
+  MeshLevel & meshLevel0 = meshBody.registerGroup< MeshLevel >( string( "Level0" ));
+  NodeManager & nodeManager = meshLevel0.getNodeManager();
+  Group & nodeSets = nodeManager.sets();
+
+  // Wellbore nodesets
+  SortedArray< localIndex > & rnegNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "rneg" ) ).reference();
+  SortedArray< localIndex > & rposNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "rpos" ) ).reference();
+  SortedArray< localIndex > & tnegNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "tneg" ) ).reference();
+  SortedArray< localIndex > & tposNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "tpos" ) ).reference();
+
+  arrayView2d< real64, nodes::REFERENCE_POSITION_USD > const & X = nodeManager.referencePosition();
+  for( int localNodeIndex=0; localNodeIndex<nodeManager.size(); ++localNodeIndex )
+  {
+    // Get Cartesian coordinates of a reference centered vertical wellbore
+    real64 & xCoord = X( localNodeIndex, 0 );
+    real64 & yCoord = X( localNodeIndex, 1 );
+    real64 const & zCoord = X( localNodeIndex, 2 );
+
+    // Compute cylindrical coordinates of a reference centered vertical wellbore
+    real64 rCoord = sqrt( xCoord * xCoord + yCoord * yCoord );
+    real64 tCoord;
+
+    if( rCoord < 1e-10 )
+    {
+      tCoord = 0.0;
+    }
+    else if( yCoord >= 0.0 )
+    {
+      tCoord = acos( xCoord/rCoord );
+    }
+    else
+    {
+      tCoord = 2 * M_PI - acos( xCoord/rCoord );
+    }
+
+    tCoord *= 180.0 / M_PI;
+
+    // Wellbore nodesets
+    if( isEqual( rCoord, m_min[0], m_positionTolerance ) )
+    {
+      rnegNodes.insert( localNodeIndex );
+    }
+
+    if( isEqual( rCoord, m_max[0], m_positionTolerance ) )
+    {
+      rposNodes.insert( localNodeIndex );
+    }
+
+    if( isEqual( tCoord, m_min[1], m_positionTolerance ) )
+    {
+      tnegNodes.insert( localNodeIndex );
+    }
+    if( isEqual( tCoord, m_max[1], m_positionTolerance ) )
+    {
+      tposNodes.insert( localNodeIndex );
+    }
+
+    // Radial distance of the outer square boundary of a reference centered vertical wellbore
+    real64 meshTheta = tCoord * M_PI / 180.0;
+    int meshAxis = static_cast< int >(round( meshTheta * 2.0 / M_PI ));
+    real64 meshPhi = fabs( meshTheta - meshAxis * M_PI / 2.0 );
+    real64 meshRout = m_max[0] / cos( meshPhi );
+
+    // Wellbore trajectory
+    real64 xTopCenter = m_trajectory[0][0];
+    real64 yTopCenter = m_trajectory[0][1];
+    real64 zTop = m_min[2];
+
+    real64 xBottomCenter = m_trajectory[1][0];
+    real64 yBottomCenter = m_trajectory[1][1];
+    real64 zBottom = m_max[2];
+
+    real64 dx = xBottomCenter - xTopCenter;
+    real64 dy = yBottomCenter - yTopCenter;
+    real64 dz = zBottom - zTop;
+    real64 dr = sqrt( dx*dx + dy*dy );
+    real64 dl = sqrt( dr*dr + dz*dz );
+
+    // Azimuth from x-axis
+    real64 theta0;
+
+    if( dr < 1e-10 )
```

Comment:
Please use the precision variable (make it `protected` if needed).

Reply:
done